### PR TITLE
feat: configurable file inclusion/exclusion for non-standard agent files

### LINKED ISSUE #287: feat: configurable file inclusion/exclusion for non-standard agent files
## Problem

Files like `DEVELOPER.md`, `AI-INSTRUCTIONS.md`, or custom naming conventions are not recognized as agent config files. Users who want to validate these files currently have no way to include them.

## Proposed Solution

Add config options to `.agnix.toml`:

```toml
# Include additional files as specific types
[files]
# Treat these patterns as ClaudeMd (gets CC-MEM, PE, etc. rules)
include_as_memory = ["DEVELOPER.md", "AI-INSTRUCTIONS.md", "CODING-GUIDELINES.md"]

# Treat these patterns as GenericMarkdown (gets XML, XP, REF rules)
include_as_generic = ["instructions/*.md"]

# Exclude files that would normally be validated
exclude = ["drafts/**"]
```

## Implementation Notes

- Extend `detect_file_type()` to check config-specified patterns before the built-in logic
- Add corresponding VS Code settings (`agnix.files.includeAsMemory`, etc.)
- LSP needs to pick up the config and pass it through `LintConfig`
- Config schema needs updating in JSON Schema for `.agnix.toml` validation

## Context

Discovered during real-world validation of 1,200+ repos - many projects use non-standard filenames for agent instructions.

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Adds a `[files]` configuration section to `.agnix.toml` that lets users override file type detection via glob patterns, enabling validation of non-standard agent instruction files.

- `include_as_memory`: glob patterns to validate files as CLAUDE.md-like instruction files (ClaudeMd rules)
- `include_as_generic`: glob patterns to validate files as generic markdown (XML, imports, cross-platform rules)
- `exclude`: glob patterns to skip files from validation entirely
- Priority order: exclude > include_as_memory > include_as_generic > built-in detection

### Key changes
- New `FilesConfig` struct in `config.rs` with TOML/JSON deserialization and schema generation
- New `resolve_file_type()` wrapper that applies config overrides on top of `detect_file_type()`
- Pre-compiled glob patterns shared via `Arc` for parallel validation performance
- LSP integration: validation, hover, and completion handlers use config-aware file type resolution
- VS Code extension settings: `agnix.files.includeAsMemory`, `agnix.files.includeAsGeneric`, `agnix.files.exclude`
- Security: path traversal validation, pattern count limits, `require_literal_separator: true`
- Localization: English, Spanish, Chinese translations for new messages

## Test Plan

- [x] 35+ new tests covering unit, integration, and edge cases
- [x] Config: TOML deserialization, defaults, invalid glob validation, path traversal rejection, pattern count limits
- [x] resolve_file_type: include/exclude mapping, priority ordering, case sensitivity, recursive globs, Windows path normalization, no-root-dir fallback
- [x] Integration: full validation pipeline with files config (include + exclude)
- [x] LSP: VsCodeFiles deserialization, merge behavior, completion provider signature update
- [x] All 1712+ existing tests continue to pass
- [x] `cargo clippy` clean on modified crates

Closes #287

### CODE REVIEW

File: crates/agnix-core/src/config.rs [hunk 1]
```diff
@@ -828,6 +881,80 @@ impl LintConfig {
...
+                            t!("core.config.files_path_traversal_suggestion").to_string(),
+                        ),
+                    });
+                }
+                // Reject absolute paths (Unix-style leading slash or Windows drive letter)
+                if normalized.starts_with('/')
+                    || (normalized.len() >= 2
+                        && normalized.as_bytes()[0].is_ascii_alphabetic()
+                        && normalized.as_bytes()[1] == b':')
+                {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs:942]: Fixed in db5e8aa - now checking for ':/' sequence to properly detect Windows absolute paths.

File: crates/agnix-core/src/lib.rs [hunk 2]
```diff
@@ -410,6 +410,137 @@ pub fn detect_file_type(path: &Path) -> FileType {
...
+                })
+            })
+            .collect()
+    };
+    Ok(CompiledFilesConfig {
+        include_as_memory: compile_include(&files.include_as_memory)?,
+        include_as_generic: compile_include(&files.include_as_generic)?,
+        exclude: compile_exclude(&files.exclude)?,
+    })
+}
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/lib.rs]: Fixed in db5e8aa - extracted compile_patterns helper function with error_builder closure parameter to eliminate duplication.

File: crates/agnix-lsp/src/vscode_config.rs [hunk 3]
```diff
@@ -170,6 +176,25 @@ pub struct VsCodeSpecs {
     pub agents_md_spec: Option<Option<String>>,
 }
 
+/// File inclusion/exclusion settings from VS Code.
+///
+/// Maps to FilesConfig in agnix-core.
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+#[serde(default, rename_all = "camelCase")]
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-lsp/src/vscode_config.rs]: Fixed in 323e124 - removed rename_all camelCase attribute so VsCodeFiles now correctly deserializes snake_case JSON keys sent by the VS Code extension. Tests updated to match actual extension behavior.

File: crates/agnix-lsp/src/backend.rs [hunk 4]
```diff
@@ -271,7 +271,7 @@ impl Backend {
         let config = Arc::clone(&*self.config.read().await);
         let registry = Arc::clone(&self.registry);
         let result = tokio::task::spawn_blocking(move || {
-            let file_type = agnix_core::detect_file_type(&file_path);
+            let file_type = agnix_core::resolve_file_type(&file_path, &config);
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-lsp/src/backend.rs:274]: Fixed in 323e124 - LSP backend now sets config.root_dir from workspace_root in both initialize() and did_change_configuration() for proper glob pattern matching.

File: crates/agnix-core/src/lib.rs [hunk 5]
```diff
@@ -527,6 +658,10 @@ pub fn validate_project_with_registry(
     // Pre-compile exclude patterns once (avoids N+1 pattern compilation)
     let exclude_patterns = compile_exclude_patterns(&config.exclude)?;
     let exclude_patterns = Arc::new(exclude_patterns);
+
+    // Pre-compile files config patterns once for the parallel walk
+    let compiled_files = Arc::new(compile_files_config(&config.files)?);
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/lib.rs]: Fixed in 3c0df30 - `validate_project_with_registry()` now treats invalid `[files]` patterns as non-fatal, falling back to an empty `CompiledFilesConfig` (via a new `Default` impl) instead of hard-failing. This aligns with `LintConfig::validate()` which only produces warnings for invalid patterns. The test was updated to expect graceful degradation.

File: crates/agnix-core/src/lib.rs [hunk 6]
```diff
@@ -578,8 +713,8 @@ pub fn validate_project_with_registry(
                 return Vec::new();
             }
 
-            // Count recognized files (detect_file_type is string-only, no I/O)
-            let file_type = detect_file_type(&file_path);
+            // Count recognized files (resolve_with_compiled is string-only, no I/O)
+            let file_type = resolve_with_compiled(&file_path, Some(&root_path), &compiled_files);
             if file_type != FileType::Unknown {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/lib.rs:715]: Fixed in 3c0df30 - extracted `validate_file_with_type()` which accepts a pre-resolved `FileType`, and the project-level parallel loop now calls it directly with the `file_type` already computed via `resolve_with_compiled()`. This eliminates the redundant per-file pattern recompilation that `validate_file_with_registry()` was doing via `resolve_file_type()`.

File: crates/agnix-lsp/src/backend.rs [hunk 7]
```diff
@@ -271,7 +271,7 @@ impl Backend {
...
             if file_type == agnix_core::FileType::Unknown {
                 return Ok(vec![]);
             }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-lsp/src/backend.rs:277]: Fixed in 323e124 - LSP backend now sets `config.root_dir` from `workspace_root` in both `initialize()` and `did_change_configuration()`, so directory-based globs like `docs/**/*.md` work correctly in the LSP.

File: crates/agnix-core/src/lib.rs [hunk 8]
```diff
@@ -410,6 +410,132 @@ pub fn detect_file_type(path: &Path) -> FileType {
...
+
+    // Compile patterns on-demand for single-file validation
+    match compile_files_config(&config.files) {
+        Ok(compiled) => resolve_with_compiled(path, config.root_dir.as_deref(), &compiled),
+        // Silent fallback is intentional: patterns are already validated at config
+        // load time (LintConfig::validate), so a compilation failure here would only
+        // occur with corrupt in-memory state. Falling back to built-in detection is
+        // the safest recovery.
+        Err(_) => detect_file_type(path),
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/lib.rs]: Fixed in 3c0df30 - the comment now accurately states that invalid patterns produce warnings during config validation but don't prevent them from reaching \, and that falling back to built-in detection is safe since the invalid patterns would have no effect anyway.

File: crates/agnix-core/src/lib.rs [hunk 9]
```diff
@@ -527,6 +653,10 @@ pub fn validate_project_with_registry(
     // Pre-compile exclude patterns once (avoids N+1 pattern compilation)
     let exclude_patterns = compile_exclude_patterns(&config.exclude)?;
     let exclude_patterns = Arc::new(exclude_patterns);
+
+    // Pre-compile files config patterns once for the parallel walk
+    let compiled_files = Arc::new(compile_files_config(&config.files)?);
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/lib.rs]: Fixed in 3c0df30 - validate_project_with_registry() now uses graceful degradation (match with fallback to CompiledFilesConfig::default()) instead of hard-failing with the ? operator. This makes the behavior consistent: LintConfig::validate() warns, resolve_file_type() falls back, and now validate_project_with_registry() also falls back.

File: docs/CONFIGURATION.md [hunk 10]
```diff
@@ -72,6 +91,7 @@ agnix automatically validates `.agnix.toml` files for:
 
 - **Invalid rule IDs**: Warns if `disabled_rules` contains IDs that don't match known patterns (AS-, CC-SK-, CC-HK-, CC-AG-, CC-MEM-, CC-PL-, XML-, MCP-, REF-, XP-, AGM-, COP-, CUR-, PE-, VER-, imports::)
 - **Unknown tools**: Warns if `tools` array contains tool names that aren't recognized
+- **Invalid file patterns**: Warns if `[files]` glob patterns have invalid syntax
 - **Deprecated fields**: Warns when using `mcp_protocol_version` (use `spec_revisions.mcp_protocol` instead)
 
 These warnings appear before validation output and include suggestions for fixes.
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply docs/CONFIGURATION.md:97]: The docs are now accurate after 3c0df30 - validate_project_with_registry() degrades gracefully on invalid patterns (falls back to empty config) instead of aborting. The docs correctly state that invalid file patterns produce warnings, which is the consistent behavior across the entire codebase.

File: crates/agnix-lsp/src/vscode_config.rs [hunk 11]
```diff
@@ -170,6 +176,25 @@ pub struct VsCodeSpecs {
...
+    #[serde(default)]
+    pub include_as_memory: Option<Vec<String>>,
+
+    /// Glob patterns for files to validate as generic markdown
+    #[serde(default)]
+    pub include_as_generic: Option<Vec<String>>,
+
+    /// Glob patterns for files to exclude from validation
+    #[serde(default)]
+    pub exclude: Option<Vec<String>>,
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-lsp/src/vscode_config.rs:195]: Fixed in 323e124 - removed rename_all camelCase attribute so VsCodeFiles now correctly deserializes the snake_case JSON keys sent by the VS Code extension.

File: crates/agnix-core/src/config.rs [hunk 12]
```diff
@@ -828,6 +881,80 @@ impl LintConfig {
...
+                        suggestion: Some(
+                            t!("core.config.files_path_traversal_suggestion").to_string(),
+                        ),
+                    });
+                }
+                // Reject absolute paths (Unix-style leading slash or Windows drive letter)
+                if normalized.starts_with('/')
+                    || (normalized.len() >= 3
+                        && normalized.as_bytes()[0].is_ascii_alphabetic()
+                        && &normalized[1..3] == ":/")
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs]: Fixed in 5f468d8 - Windows path check now uses \ to avoid potential UTF-8 boundary panic.

[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs]: Fixed in 5f468d8 - now uses bytes API to safely check Windows drive letter prefix.

File: crates/agnix-core/src/lib.rs [hunk 13]
```diff
@@ -410,6 +410,132 @@ pub fn detect_file_type(path: &Path) -> FileType {
...
+}
+
+/// Match options for file inclusion/exclusion glob patterns.
+///
+/// `require_literal_separator` is `true` so that `*` only matches within a
+/// single path component. Users must use `**` for recursive matching (e.g.
+/// `dir/**/*.md` instead of `dir/*.md` to match nested files).
+const FILES_MATCH_OPTIONS: glob::MatchOptions = glob::MatchOptions {
+    case_sensitive: true,
+    require_literal_separator: true,
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/lib.rs]: Fixed in 5f468d8 - compile_files_config now uses lenient compilation that silently skips invalid patterns instead of failing all. Valid patterns still work even when others are invalid.
